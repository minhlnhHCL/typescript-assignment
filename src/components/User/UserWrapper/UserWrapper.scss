.user-card {
    padding: 16px;
    border-radius: 8px;
    box-shadow: rgba(0, 0, 0, 0.15) 1.95px 1.95px 2.6px;
    margin: 16px 0px;

    .user-wrapper {
        .avatar {
            height: 120px;
            width: 120px;
            background: rgb(76, 117, 161);
            border-radius: 50%;

            img {
                width: 100%;
                height: 100%;
                object-fit: cover;
            }
        }

        .info {
            display: flex;
        }

        .btn {
            border-radius: 8px;
            padding: 8px;
            border-color: transparent;
            width: 100%;
            cursor: pointer;
            color: #fff;

            &.edit {
                background: rgb(113, 158, 183);
            }
            &.cancel {
                background: rgba(228, 79, 79, 0.885);
            }
        }
    }
}