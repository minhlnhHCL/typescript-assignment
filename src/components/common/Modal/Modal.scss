.d-flex-ac {
    display: flex;
    align-items: center;
}

.d-flex-sb {
    display: flex;
    justify-content: space-between;
    align-items: center;
}

.mt-32 {
    margin-top: 32px;
}

.modal-actions {
    position: absolute;
    bottom: 0px;
    left: 0px;
    width: 100%;
    padding: 18px 0px;
    background-color: rgba(243, 243, 243, 0.2);
    border-radius: 0px 0px 8px 8px;

    .action-container {
        width: 100%;
        display: flex;
        justify-content: flex-end;

        .btn {
            border-radius: 8px;
            padding: 8px;
            border-color: transparent;
            cursor: pointer;
            letter-spacing: 1.5px;

            &.cancel {
                margin-right: 8px;
                background: transparent;
            }

            &.submit {
                width: 120px;
                margin-right: 24px;
                background: rgb(103, 168, 203);
                color: #fff;

                &:disabled {
                    border: 1px solid #999999;
                    background-color: #cccccc;
                    color: #666666;
                }
            }
        }
    }
}

.modal-container {
    position: fixed;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    transform: scale(0);
    z-index: 1;
    display: flex;

    .modal-background {
        background: rgba(0, 0, 0, .8);
        width: 100%;
        height: 100%;
        display: flex;
        justify-content: center;
        align-items: center;

        .modal {
            background: white;
            width: 320px;
            height: 220px;
            display: flex;
            flex-direction: column;
            justify-content: center;
            border-radius: 8px;
            padding: 24px;
            position: relative;

            .modal-header {
                h3 {
                    margin: 0px;
                }

                .btn-close {
                    background: rgba(228, 79, 79, 0.885);
                    color: #fff;
                    border: 0;
                    border-radius: 50%;
                    cursor: pointer;
                    width: 35px;
                    height: 35px;
                    padding: 4px;
                    font-size: 1rem;
                }
            }

            .modal-content {
                margin-top: 32px;
                height: 100%;
                width: 100%;
            }
        }
    }

    &.active {
        z-index: 0;
        transform: scale(1);

        .modal-background {
            background: rgba(0, 0, 0, .0);
            animation: fadeIn .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;

            .modal {
                transform: translateX(-1500px);
                animation: roadRunnerIn .3s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
            }
        }
    }

    &.out {
        animation: quickScaleDown 0s .5s linear forwards;
        opacity: 0;
        visibility: hidden;
        transition: opacity 1.5s, visibility 1.5s;

        .modal-background {
            animation: fadeOut .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;

            .modal {
                animation: roadRunnerOut .5s cubic-bezier(0.165, 0.840, 0.440, 1.000) forwards;
            }
        }
    }
}

@keyframes fadeIn {
    0% {
        background: rgba(0, 0, 0, .0);
    }
    100% {
        background: rgba(0, 0, 0, .7);
    }
}

@keyframes fadeOut {
    0% {
        background: rgba(0, 0, 0, .7);
    }

    100% {
        background: rgba(0, 0, 0, .0);
    }
}

@keyframes quickScaleDown {
    0% {
        transform: scale(1);
    }

    99.9% {
        transform: scale(1);
    }

    100% {
        transform: scale(0);
    }
}


@keyframes roadRunnerIn {
    0% {
        transform: translateX(-1500px) skewX(30deg) scaleX(1.3);
    }

    70% {
        transform: translateX(30px) skewX(0deg) scaleX(.9);
    }

    100% {
        transform: translateX(0px) skewX(0deg) scaleX(1);
    }
}

@keyframes roadRunnerOut {
    0% {
        transform: translateX(0px) skewX(0deg) scaleX(1);
    }

    30% {
        transform: translateX(-30px) skewX(-5deg) scaleX(.9);
    }

    100% {
        transform: translateX(1500px) skewX(30deg) scaleX(1.3);
    }
}