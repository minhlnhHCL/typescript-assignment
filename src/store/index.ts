import { configureStore, PreloadedState } from '@reduxjs/toolkit'
import { useDispatch, useSelector } from 'react-redux'
import { combineReducers } from 'redux'

import users from './userSlice'
import modal from './modalSlice'
import toast from './toastSlice'

const reducer = combineReducers({
    users,
    modal,
    toast
})
const setupStore = (preloadedState?: PreloadedState<RootState>) => {
    return configureStore({
        reducer,
        preloadedState,
        middleware: (getDefaultMiddleware) =>
            getDefaultMiddleware({
                serializableCheck: false
            })
    })
}

export type RootState = ReturnType<typeof reducer>
export type AppStore = ReturnType<typeof setupStore>
export type AppDispatch = AppStore['dispatch']

export default setupStore
