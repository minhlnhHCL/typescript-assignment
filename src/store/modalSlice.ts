import { createSlice, PayloadAction } from "@reduxjs/toolkit"


export interface IModal {
    open: boolean,
    title: string,
    children: JSX.Element | string | null
    successCb: () => void

}

const initialState: IModal = {
    open: false,
    title: '',
    children: null,
    successCb: () => { },
}

export const modalSlice = createSlice({
    name: 'modal',
    initialState,
    reducers: {
        setModal: (state, action: PayloadAction<IModal>) => {
            state.children = action.payload.children
            state.open = action.payload.open
            state.title = action.payload.title
            state.successCb = action.payload.successCb
        }
    }
})

export const { setModal } = modalSlice.actions
export default modalSlice.reducer