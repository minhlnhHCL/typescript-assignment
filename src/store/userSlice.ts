import { createSlice, PayloadAction } from '@reduxjs/toolkit'
import { IUser } from '../components/User/UserInterface'
export const GET_USER = 'GET_USER'
export const ADD_USER = 'ADD_USER'
export const EDIT_USER = 'EDIT_USER'
export const DELETE_USER = 'DELETE_USER'
export const CURRENT_USER = 'CURRENT_USER'

export interface UserState {
    users: IUser[],
}

export const initialState: UserState = {
    users: [],
}
//UserSlice

export const userSLice = createSlice({
    name: 'users',
    initialState,
    reducers: {
        getUsers: (state, action: PayloadAction<IUser[]>) => {
            if (state.users.length == 1) {
                const newUserArr = [...state.users, ...action.payload]
                state.users = newUserArr
                return state
            }
            state.users = action.payload
        },
        addUser: (state, action: PayloadAction<IUser>) => {
            const index = state.users.findIndex(el => el.id == action.payload.id);
            if (index == -1) {
                state.users = [action.payload, ...state.users]
                return state
            } else {
                const newPayload = { ...action.payload, id: action.payload.id && action.payload.id + 1 }
                state.users = [newPayload, ...state.users]
                return state
            }
        },
        editUser: (state, action: PayloadAction<IUser>) => {
            const index = state.users.findIndex(el => el.id == action.payload.id);
            const newUserArr = [...state.users]
            newUserArr[index] = action.payload
            state.users = newUserArr
        },
        deleteUser: (state, action: PayloadAction<IUser>) => {
            const clonedArr = [...state.users]
            const newUserArr = clonedArr.filter(item => item.id !== action.payload.id)
            state.users = newUserArr
        }
    }
})


export const { getUsers, editUser, addUser, deleteUser } = userSLice.actions
export default userSLice.reducer