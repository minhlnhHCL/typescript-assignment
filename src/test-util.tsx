import React, { PropsWithChildren, ReactNode } from 'react'
import { render } from '@testing-library/react'
import type { RenderOptions } from '@testing-library/react'
import type { PreloadedState } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'

import setupStore from '/store'
import type { AppStore, RootState } from '/store/index'
import { LoadingProvider } from '/context/LoadingContext'

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export * from '@testing-library/react'

export const renderWithRedux = (
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) => {
  const Wrapper = ({ children }: PropsWithChildren<{}>): JSX.Element => {
    return <Provider store={store}>
      <LoadingProvider>
        {children}
      </LoadingProvider>
    </Provider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}