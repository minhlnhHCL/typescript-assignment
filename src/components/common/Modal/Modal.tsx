import './Modal.scss'
import { useNavigate } from 'react-router-dom'
import { useAppDispatch, useAppSelector } from 'store/hooks'
import PortalContainer from '../Portal/Portal'
import { setModal } from 'store/modalSlice'
import { createPortal } from 'react-dom'

const Modal = () => {
  const dispatch = useAppDispatch()
  const { children, open, title, successCb } = useAppSelector(state => state.modal)

  const closeHandler = () => {
    dispatch(setModal({
      open: false,
      title: '',
      children: '',
      successCb: () => { }
    }))
  }

  const content = (
    <div className={`modal-container ${open ? 'active' : 'active out'}`}>
      <div className='modal-background'>
        <div className="modal">
          <div className="modal-header">
            <div className='d-flex-sb'>
              <h3>{title}</h3>
              <button className='btn-close' onClick={closeHandler}>X</button>
            </div>
          </div>
          <div className='modal-content'>
            {children}
          </div>
          <div className="modal-actions">
            <div className="action-container">
              <button className='btn cancel' onClick={closeHandler}>cancel</button>
              <button className='btn submit' onClick={successCb}>OK</button>
            </div>
          </div>
        </div>
      </div>
    </div>

  )
  return createPortal(content, document.getElementById('modal-root') as HTMLElement);
}

export default Modal