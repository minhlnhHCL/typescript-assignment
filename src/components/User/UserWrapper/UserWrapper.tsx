import './UserWrapper.scss'
import { useNavigate } from 'react-router-dom'
import { deleteUser } from '/store/userSlice'
import { useAppDispatch } from '/store/hooks'
import { deleteReq } from '/utils/api-requests'
import { IUser } from '../UserInterface'
import { useLoadingContext } from '/context/LoadingContext'
import { setModal } from '/store/modalSlice'

const UserWrapper = ({ data }: { data: IUser }) => {
    const dispatch = useAppDispatch()
    const navigate = useNavigate()
    const { offLoading, onLoading } = useLoadingContext()
    const deleteHandler = async () => {
        try {
            onLoading()
            const req = await deleteReq({ url: `/${data.id}` })
            if (req.status == 200) {
                return dispatch(deleteUser(data))
            }
        } catch (error) {
            console.log(error)
        } finally {
            offLoading()
            dispatch(setModal({
                open: false,
                title: '',
                children: '',
                successCb: () => { }
            }))
        }
    }

    const onDelete = () => {
        dispatch(setModal({
            title: 'Confirmation',
            children: 'Are you sure you want to delete',
            open: true,
            successCb: deleteHandler
        }))
    }

    return (
        <div className='user-card'>
            <div className='user-wrapper d-flex-sb'>
                <div className='d-flex'>
                    <div className='avatar'><img src={data.image} /></div>
                    <div className='info ml-32 d-fc'>
                        <div className='d-flex-wrap'>
                            <p>{data.firstName}</p>
                            <p className='ml-4'>{data.lastName}</p>
                        </div>
                        <div className='mt--20 fs-14'>
                            <p>{data.company?.title}</p>
                        </div>
                    </div>
                </div>
                <div>
                    <button className='btn edit' onClick={() => {
                        navigate(`/edit/${data.id}`)
                    }}>Edit</button>
                    <button className='btn cancel mt-8' onClick={onDelete}>
                        Delete
                    </button>
                </div>
            </div>
        </div>
    )
}

export default UserWrapper
