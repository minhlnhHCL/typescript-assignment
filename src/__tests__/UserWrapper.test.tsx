import { fireEvent, render, screen, waitFor } from "@testing-library/react"
import '@testing-library/jest-dom'
import { IUser } from "components/User/UserInterface"
import UserWrapper from "components/User/UserWrapper/UserWrapper"
import { useNavigate } from "react-router-dom"
import { useAppDispatch, useAppSelector } from "store/hooks"
import { renderWithRedux } from "test-util"
import { deleteReq } from "utils/api-requests"
import Modal from 'components/common/Modal/Modal'
import ReactDOM from "react-dom"
import { ReactNode, ReactPortal } from "react"
import { setModal } from "store/modalSlice"
import { deleteUser } from "store/userSlice"

jest.mock("store/hooks", () => ({
    ...jest.requireActual("react-redux"),
    useAppSelector: jest.fn(),
    useAppDispatch: jest.fn(),
}))

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: jest.fn(),
    useParams: jest.fn()
}))

jest.mock('utils/api-requests')

const editUserData: IUser = {
    id: 1,
    firstName: "Minh",
    lastName: "Luong",
    image: 'https://robohash.org/hicveldicta.png',
    age: 25,
    company: {
        title: "Front End Engineer",
    },
}

describe('User Wrapper', () => {
    const oldCreatePortal = ReactDOM.createPortal;
    beforeAll(() => {
        ReactDOM.createPortal = (node: ReactNode): ReactPortal =>
            node as ReactPortal;
    });

    afterAll(() => {
        ReactDOM.createPortal = oldCreatePortal;
    });
    beforeEach(() => {
        (useAppDispatch as jest.Mock).mockImplementationOnce(() => jest.fn());
        (useNavigate as jest.Mock).mockReturnValue(() => { });
        (useAppSelector as jest.Mock).mockImplementation(() => { });
    });

    it('should render user info', async () => {
        const { getByText, container } = renderWithRedux(<UserWrapper data={editUserData} />);
        expect(getByText('Minh')).toBeInTheDocument();
        expect(getByText('Luong')).toBeInTheDocument();
        expect(getByText('Front End Engineer')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    })


    it('should delete selected user', async () => {
        const onLoading = jest.fn();
        const offLoading = jest.fn();
        const successCb = jest.fn();
        (deleteReq as jest.Mock).mockResolvedValue({
            data: editUserData
        });
        (useAppSelector as jest.Mock).mockResolvedValue({
            modal: {
                title: 'Confirmation',
                children: 'Are you sure you want to delete this user ?',
                open: true,
                successCb: successCb
            }
        })
        const { store } = renderWithRedux(<UserWrapper data={editUserData} />, { onLoading, offLoading });
        fireEvent.click(screen.getByRole('button', { name: 'Delete' }));
        expect(useAppDispatch).toBeCalled();
        render(<Modal />)
        fireEvent.click(screen.getByRole('button', { name: 'OK' }));
        expect(useAppDispatch).toBeCalled();
        store.dispatch(setModal({
            title: 'Confirmation',
            children: 'Are you sure you want to delete this user ?',
            open: true,
            successCb: successCb
        }));
        store.dispatch(deleteUser(editUserData));
        expect(store.getState().users.users).toEqual([])
    });
})