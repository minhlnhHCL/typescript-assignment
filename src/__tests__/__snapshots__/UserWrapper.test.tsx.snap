// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`User Wrapper should render user info 1`] = `
<div>
  <div
    class="user-card"
  >
    <div
      class="user-wrapper d-flex-sb"
    >
      <div
        class="d-flex"
      >
        <div
          class="avatar"
        >
          <img
            src="https://robohash.org/hicveldicta.png"
          />
        </div>
        <div
          class="info ml-32 d-fc"
        >
          <div
            class="d-flex-wrap"
          >
            <p>
              Minh
            </p>
            <p
              class="ml-4"
            >
              Luong
            </p>
          </div>
          <div
            class="mt--20 fs-14"
          >
            <p>
              Front End Engineer
            </p>
          </div>
        </div>
      </div>
      <div>
        <button
          class="btn edit"
        >
          Edit
        </button>
        <button
          class="btn cancel mt-8"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</div>
`;
